compiler/rustc_mir/src/transform/const_prop.rs

```diff
@@ -385,15 +385,19 @@ impl<'mir, 'tcx> ConstPropagator<'mir, 'tcx> {
             (),
         );
 
-        let ret = ecx
-            .layout_of(body.return_ty().subst(tcx, substs))
-            .ok()
+        let ret = if let Ok(layout) = ecx.layout_of(body.return_ty().subst(tcx, substs)) {
             // Don't bother allocating memory for ZST types which have no values
             // or for large values.
-            .filter(|ret_layout| {
-                !ret_layout.is_zst() && ret_layout.size < Size::from_bytes(MAX_ALLOC_LIMIT)
-            })
-            .map(|ret_layout| ecx.allocate(ret_layout, MemoryKind::Stack).into());
+            if !layout.is_zst() && layout.size < Size::from_bytes(MAX_ALLOC_LIMIT) {
+                // hopefully all types will allocate, since large types have already been removed,
+                // but check anyways
+                ecx.allocate(layout, MemoryKind::Stack).ok().map(Into::into)
```

Comment:
Am I reading this correctly, that the only reason you unpack these combinators is that `allocate` can now fail, which would give you an `Option<Option<_>>` after the `map`?
If so, it would be simpler to replace `map` with `and_then`:
```rust
.and_then(|ret_layout| ecx.allocate(ret_layout, MemoryKind::Stack).ok())
.map(Into::into);
```